### PR TITLE
RE-1839 Build Summary Date Accounting Errors

### PR DESCRIPTION
**Pull request by hughsaunders**
Addresses two grave bugs in the way build summary handles dates when
    producing graphs:

    1) The graphs are generated using a histogram, builds are divided into 60
    24 hour buckets, and each bucket is labeled with a date. The problem is
    that the start point is whenever the page is rendered so the 24 hour
    histogram buckets do not line up with calendar days. The resolution is to
    move the histogram start point to a day boundary so that histogram buckets
    and days lineup.

    2) The way that builds are allocated to buckets is by working out the
    difference between the histogram end, and the build timestamp in
    milliseconds, this in the divided by the number of milliseconds in a day
    and rounded to produce an integer. That integer is subtracted from the
    length of the histogram to get the bucket number.  Due to standard rounding
    some builds would be put in the correct bucket, and some in the bucket for
    the next day. The fix is to use the ceil function instead of round.

Issue: [RE-1839](https://rpc-openstack.atlassian.net/browse/RE-1839)

### CODE REVIEW

**Comment by hughsaunders** [comment]
recheck_unit_tests failed on apt.

**Comment by hughsaunders** [comment]
Updated beta page with changes applied, data is from Friday so most recent few days are missing. http://rpc-repo.rackspace.com/rpcgating/buildsummarybeta/index.html#/

File: scripts/build_summary/index.html [hunk 1]
```diff
@@ -164,28 +164,48 @@ <h1 v-else>Loading all of the JSON... please wait</h1>
...
+          return end
+        },
         labels: function(){
-          now = new Date()
           return [...Array(this.retention_days).keys()].map(
             (i) => {
               var daysAgo = this.retention_days-i
               var dayInMillis = 86400000
-              return new Date(now-dayInMillis * daysAgo).toLocaleDateString()
+              return new Date(this.histogram_end-dayInMillis * daysAgo).toLocaleDateString()
```
Comment (odyssey4me) [inline scripts/build_summary/index.html]: nit: It'd be nice to see spaces around the operators here. I got confused initially, thinking that ``this.histogram_end-dayInMillis`` was a variable name.

File: scripts/build_summary/index.html [hunk 2]
```diff
@@ -164,28 +164,48 @@ <h1 v-else>Loading all of the JSON... please wait</h1>
...
+          // up with the date labels on the graphs. It also meant that the values for
+          // each date varied during the day as a build could age enough to move up to
+          // the next bucket. This was exacerbated by a rounding error that caused
+          // builds for one day to be evenly split over two buckets
+          // (the correct day and the next day)
+          // Now histogram end is aligned with a day boundary so that the 24 hour
+          // buckets are aligned with calendar days.
+          end = new Date() // now
+          end.setHours(0,0,0,0) //start of today
+          end.setDate(end.getDate()+1) //start of tomorrow
```
Comment (odyssey4me) [inline scripts/build_summary/index.html]: nit: It'd be nice to see spaces around the operators here.

File: scripts/build_summary/index.html [hunk 3]
```diff
@@ -164,28 +164,48 @@ <h1 v-else>Loading all of the JSON... please wait</h1>
...
       },
       methods: {
         histogram: function(builds, length, inc) {
           histogram = new Array(length).fill(0);
-          now = new Date()
-          vueobj = this
+          histogram_end = this.histogram_end
+          // oldest bucket
+          histogram_start = new Date(histogram_end)
+          histogram_start.setDate(histogram_start.getDate()-length)
```
Comment (odyssey4me) [inline scripts/build_summary/index.html]: nit: It'd be nice to see spaces around the operators here.

**Comment by hughsaunders** [comment]
Added another commit to address whitespace errors pointed out by @odyssey4me.

File: scripts/build_summary/cachequery.py [hunk 4]
```diff
@@ -1,23 +1,33 @@
-import click
-import pickle
+import json
+import dateutil.parser
 
 
-@click.group()
-def cli():
-    pass
+def loadbd(filename):
```
[the remark this answers is not in the document]
Comment (hughsaunders) [reply scripts/build_summary/cachequery.py]: They are just useful functions for working with the cache file. The CLI tool was not useful at all as it only worked with the now unused pickle dump. I just source the file in ipython and inspect the data.

[the remark this answers is not in the document]
Comment (hughsaunders) [reply scripts/build_summary/cachequery.py]: correct "builddata"

File: scripts/build_summary/failure.py [hunk 5]
```diff
@@ -447,11 +445,3 @@ def scan(self):
                     self.matches = True
                     self.detail = str.strip()
                     break
-
-
-class UnknownFailure(Failure):
```
[the remark this answers is not in the document]
Comment (hughsaunders) [reply scripts/build_summary/failure.py]: Builds that failed with 0 known failures attached. In a previous iteration of build summary, build failure was determined by the number of failures attached, so all failed builds had to have a failure. Now each build has a failed boolean. Also successful builds can have failures (eg for retries). So the number of failures is no longer the indicator of overall build success.

File: scripts/build_summary/index.html [hunk 6]
```diff
@@ -201,9 +202,8 @@ <h1 v-else>Loading all of the JSON... please wait</h1>
                   newest: dsb.slice(-1)[0]}
         },
         dfmt: function(date_string) {
```
[the remark this answers is not in the document]
Comment (hughsaunders) [reply scripts/build_summary/index.html]: Done

File: scripts/build_summary/cachequery.py [hunk 7]
```diff
@@ -1,23 +1,33 @@
...
+    data = json.load(open(filename))
+    builds = data['builds']
+    failures = data['failures']
 
+    # transform the uuids used to reference objects in json, back into
+    # object references
+    # also convert iso 8601 timestamps into date objects.
+    for b in builds.values():
```
[the remark this answers is not in the document]
Comment (hughsaunders) [reply scripts/build_summary/cachequery.py]: Used the first one, thanks.

File: scripts/build_summary/cachequery.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hughsaunders) [reply scripts/build_summary/cachequery.py]: Added a comment

[the remark this answers is not in the document]
Comment (hughsaunders) [reply scripts/build_summary/cachequery.py]: Added a comment about the purpose of the file

**Pull request merged**